### PR TITLE
SI-8059 Override immutable.Queue#{+:,:+} for performance

### PR DESCRIPTION
**Pull request by soc**
Without those overrides, all elements are unnecessarily copied.

### CODE REVIEW

**Comment by soc** [comment]
Is this what you had in mind, @kimhanse?

**Comment by kimhanse** [comment]
Yes, that looks like it would solve my problem

**Comment by soc** [comment]
Review by @Ichoran, please. Added some tests.

File: src/library/scala/collection/immutable/Queue.scala [hunk 1]
```diff
@@ -89,6 +89,12 @@ class Queue[+A] protected(protected val in: List[A], protected val out: List[A])
    */
   override def length = in.length + out.length
 
+  override def +:[B >: A, That](elem: B)(implicit bf: CanBuildFrom[Queue[A], B, That]): That =
+    new Queue(in, elem :: out).asInstanceOf[That]
+
+  override def :+[B >: A, That](elem: B)(implicit bf: CanBuildFrom[Queue[A], B, That]): That =
+    enqueue(elem).asInstanceOf[That]
```
Comment (soc) [inline src/library/scala/collection/immutable/Queue.scala]: I wonder if there is a better way than casting ...

Comment (retronym) [inline src/library/scala/collection/immutable/Queue.scala]: The cast tells you that you're ignoring the provided builder.

See the implementation in `List`:

```
override def +:[B >: A, That](elem: B)(implicit bf: CanBuildFrom[List[A], B, That]): That = bf match {
    case _: List.GenericCanBuildFrom[_] => (elem :: this).asInstanceOf[That]
    case _ => super.+:(elem)(bf)
  }
```

We still have to cast, but the semantics are correct.

**Comment by Ichoran** [comment]
LGTM.  I'm pretty sure that there are better implementations of take and drop waiting out there, but if nobody cares that much, maybe it's not worth spending the time on it now.

**Pull request merged**